### PR TITLE
fix: font selection regression

### PR DESCRIPTION
**Pull request by andrew-spare**
This recent change https://github.com/diegomura/react-pdf/pull/2640 appears to have caused a regression where the first font used in a line of text is utilized for the entire line. This includes usage of `fontFamily`, `fontWeight`, and `fontStyle`.

# Changes
* Fixed `pickFontFromFontStack` to select from `fontStack` before `lastFont`. This should now properly prioritize fonts
* Added unit tests: `packages/layout/tests/text/fontSubstitution.test.js`
* Updated example document to include this scenario

# Related

* https://github.com/diegomura/react-pdf/pull/2640
* https://github.com/diegomura/react-pdf/issues/2730
* https://github.com/wojtekmaj/react-pdf/issues/1790

# Example
```
<Text style={{ fontSize: 10 }}>
  The following are multiple font families, weights, and styles all on the
  same line
</Text>
<Text style={{ fontFamily: 'Roboto' }}>
  Roboto Normal{' / '}
  <Text style={{ fontWeight: 'bold' }}>Roboto Bold</Text>
  {' / '}
  <Text style={{ fontStyle: 'italic' }}>Roboto Italic</Text>
  {' / '}
  <Text style={{ fontFamily: 'Courier' }}>Courier</Text>
</Text>
```

## Before
example-before

## After
example-after

### CODE REVIEW

**Comment by iSilkline** [comment]
@diegomura could you please merge this?

**Comment by TomasSalas** [comment]
@diegomura Plase merge this ! i need :(

**Comment by websnacks** [comment]
Yes I am using this library as well and right now the fonts are not working correctly. This fix is what I am waiting for! Please merge!

**Comment by TomasSalas** [comment]
@websnacks Dear, when can we see the change reflected in the library? I am waiting for this change to upload a code to improve a PDF

**Comment by TomasSalas** [comment]
When are the changes going to go up? I need it @diegomura @bdkopen

**Comment by Hatko** [comment]
I applied these changes, but still couldn't get the bold font back. Instead, to resolve it, I forced the resolution of "@react-pdf/layout" to point to `3.11.5`. We are using `bun`, for it you need to add:

```
"overrides": {
    "@react-pdf/layout": "3.11.5"
},
```

to your `package.json`

**Comment by TomasSalas** [comment]
@Hatko What you tell me doesn't work for me since I work with @react-pdf/renderer

I have not implemented @react-pdf/layout in my project

Captura de pantalla 2024-05-23 a la(s) 9 43 01 a  m
Captura de pantalla 2024-05-23 a la(s) 9 43 19 a  m
Captura de pantalla 2024-05-23 a la(s) 9 43 54 a  m



This is where I have the problem

**Comment by lecstor** [comment]
@TomasSalas the fix needs to be applied to the `@react-pdf/layout` package which `@react-pdf/renderer` depends on.
I applied the fix using `yarn patch @react-pdf/layout` (as I use yarn for my package manager) and can confirm that it resolved the issue in our application.

**Comment by TomasSalas** [comment]
@lecstor Could you help me with my project? since I work with NPM and I do not fully understand what you are telling me to be able to use what I need in the texts ??

**Comment by lecstor** [comment]
@TomasSalas it looks like you could use this then.. https://www.npmjs.com/package/patch-package

You'll need to open `node_modules/@react-pdf/layout/lib/index.js` in your project and apply the fix in this PR. (and do the same to `node_modules/@react-pdf/layout/lib/index.cjs`)

Then follow the instructions in `patch-package` to create the patch and commit it to your repo.

Screenshot 2024-05-24 at 7 58 39 AM

**Comment by TomasSalas** [comment]
@lecstor 
Captura de pantalla 2024-05-23 a la(s) 6 09 44 p  m

I made the change in node modules and nothing changes

Captura de pantalla 2024-05-23 a la(s) 9 46 15 a  m

**Comment by lecstor** [comment]
@TomasSalas have a look at the PR. Did you remove lines 204 to 206?

**Comment by TomasSalas** [comment]
@lecstor yes remove lines 204 to 206

[
Captura de pantalla 2024-05-23 a la(s) 6 09 44 p  m
](url)

**Comment by lecstor** [comment]
@TomasSalas not according to that screenshot in your last comment.
Please look at the PR changes.

These lines should not remain..

```
if (lastFont) {
  fontStackWithFallback.unshift(lastFont);
}
```

Screenshot 2024-05-24 at 8 21 44 AM

**Comment by TomasSalas** [comment]
@lecstor 
It worked ! Thank you very much.. the last question, if this goes to production, will there be any problems? Or what should I do when building the app?

**Comment by lecstor** [comment]
@TomasSalas as long as you add the `postinstall` script as per the instructions the package should be patched whenever `npm install` is run, which should include during CI when deploying to prod.

**Comment by JustJoostNL** [comment]
I have applied the patch like above, but unfortunately it's still not working (font weight).

**Comment by TomasSalas** [comment]
Is it known when the changes will be uploaded? @diegomura @bdkopen

**Comment by bdkopen** [comment]
> Is it known when the changes will be uploaded? @diegomura @bdkopen

I am not a maintainer, so I have no control over when this will be merged.

**Comment by TomasSalas** [comment]
@bdkopen And do you know who the maintainer is? so I can request his help?

**Comment by bdkopen** [comment]
> @bdkopen And do you know who the maintainer is? so I can request his help?
> 

@diegomura is the primary maintainer.

**Comment by C4T4** [comment]
@diegomura merge needed

**Comment by TomasSalas** [comment]
@diegomura please 🙏🏻

**Comment by iSilkline** [comment]
@diegomura It's been 2+ months in the last merge. Are you still able to maintain this repo?

**Review by diegomura — APPROVED** [review]
utACK

**Comment by TomasSalas** [comment]
I just updated the library and it doesn't work

**Comment by iSilkline** [comment]
@TomasSalas it's because it's not released yet. last release is april 25

**Comment by TomasSalas** [comment]
When can we have this in production, I need this change urgently :( @diegomura

**Comment by TomasSalas** [comment]
After the update the following happens
if I use the Text tag inside another Text tag it takes the style of the inner tag and not just what that tag wraps

image

image


Help Please :(

**Pull request merged**
